app/domain/authentication/authn_oidc/authenticator.rb

```diff
@@ -0,0 +1,84 @@
+module Authentication
+  module AuthnOidc
+    class AuthenticationError < RuntimeError; end
+    class NotFoundError < RuntimeError; end
+    class OIDCConfigurationError < RuntimeError; end
+    class OIDCAuthenticationError < RuntimeError; end
+
+    class Authenticator
+
+      def initialize(env:)
+        @env = env
+      end
+
+      def valid?(input)
+        @authenticator_name = input.authenticator_name
+        @service_id = input.service_id
+        @conjur_account = input.account
+        @request_body = input.request.body.read
+
+        verify_service_enabled
+
+        oidc_authn_service = AuthenticationService.new(service.identifier, conjur_account)
+        id_token, user_info = oidc_authn_service.get_user_details(request_body)
+
+        # TODO: validate id_token - if not raise error
+
+        validate_user_info(user_info, id_token.sub)
+
+        username = user_info.preferred_username
+        input.instance_variable_set(:@username, username)
```

Comment:
`input` is intentionally immutable, and breaking the immutability with this backchannel method seriously undermines the design.

I'll discuss alternative solutions to the larger problem this is being used to solved in other comments or in my review at the end.

Reply:
This was done so that we can use the already written code such as `validate_security` etc. which gets the username from the input. If this isn't good we need to find a way to change the design a bit in other authenticators too.

Reply:
Let's discuss this on Tuesday.  I may have done some refactoring of this by then.